core/vm/contracts_write.go

```diff
@@ -0,0 +1,161 @@
+package vm
+
+import (
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/harmony-one/harmony/shard"
+	staking "github.com/harmony-one/harmony/staking"
+	stakingTypes "github.com/harmony-one/harmony/staking/types"
+)
+
+// WriteCapablePrecompiledContractsStaking lists out the write capable precompiled contracts
+// which are available after the StakingPrecompileEpoch
+// for now, we have only one contract at 252 or 0xfc - which is the staking precompile
+var WriteCapablePrecompiledContractsStaking = map[common.Address]WriteCapablePrecompiledContract{
+	common.BytesToAddress([]byte{252}): &stakingPrecompile{},
+}
+
+// WriteCapablePrecompiledContract represents the interface for Native Go contracts
+// which are available as a precompile in the EVM
+// As with (read-only) PrecompiledContracts, these need a RequiredGas function
+// Note that these contracts have the capability to alter the state
+// while those in contracts.go do not
+type WriteCapablePrecompiledContract interface {
+	// RequiredGas calculates the contract gas use
+	RequiredGas(evm *EVM, input []byte) uint64
+	// use a different name from read-only contracts to be safe
+	RunWriteCapable(evm *EVM, contract *Contract, input []byte) ([]byte, error)
+}
+
+// RunWriteCapablePrecompiledContract runs and evaluates the output of a write capable precompiled contract.
+func RunWriteCapablePrecompiledContract(p WriteCapablePrecompiledContract, evm *EVM, contract *Contract, input []byte, readOnly bool) ([]byte, error) {
+	// immediately error out if readOnly
+	if readOnly {
+		return nil, errWriteProtection
+	}
+	gas := p.RequiredGas(evm, input)
+	if !contract.UseGas(gas) {
+		return nil, ErrOutOfGas
+	}
+	return p.RunWriteCapable(evm, contract, input)
+}
+
+type stakingPrecompile struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *stakingPrecompile) RequiredGas(evm *EVM, input []byte) uint64 {
+	// If you check the tests, they contain a bunch of successful transactions
+	// which are of the type Delegate / Undelegate / CollectRewards
+	// these consume anywhere between 22-25k of gas, which is in line with
+	// what would happen for a non-EVM staking tx
+	// so no additional gas needs to be consumed here
+	return 0
```

Comment:
Not sure I got the comment. Why do we charge 0 gas for calling these staking compiles?